go/vt/vtgate/buffer/buffer.go

```diff
@@ -67,12 +69,61 @@ const (
 // currently retried.
 type RetryDoneFunc context.CancelFunc
 
+const (
+	ClusterEventReshardingInProgress = "current keyspace is being resharded"
+	ClusterEventReparentInProgress   = "primary is not serving, there may be a reparent operation in progress"
+	ClusterEventMoveTables           = "disallowed due to rule"
+)
+
+var ClusterEvents []string
+
+func init() {
+	ClusterEvents = []string{
+		ClusterEventReshardingInProgress,
+		ClusterEventReparentInProgress,
+		ClusterEventMoveTables,
+	}
+}
+
 // CausedByFailover returns true if "err" was supposedly caused by a failover.
 // To simplify things, we've merged the detection for different MySQL flavors
 // in one function. Supported flavors: MariaDB, MySQL
 func CausedByFailover(err error) bool {
 	log.V(2).Infof("Checking error (type: %T) if it is caused by a failover. err: %v", err, err)
-	return vterrors.Code(err) == vtrpcpb.Code_CLUSTER_EVENT
+	reason, isFailover := isFailoverError(err)
+	if isFailover {
+		log.Infof("CausedByFailover signalling failover for reason: %s", reason)
+	}
+	return isFailover
+}
+
+// for debugging purposes
+func getReason(err error) string {
+	for _, ce := range ClusterEvents {
+		if strings.Contains(err.Error(), ce) {
+			return ce
+		}
+	}
+	return ""
+}
+
+// isFailoverError looks at the error returned by the sql query execution to check if there is a cluster event
+// (caused by resharding or reparenting) or a denied tables error seen during switch writes in MoveTables
+func isFailoverError(err error) (string, bool) {
+	var reason string
+	var isFailover bool
+	switch vterrors.Code(err) {
+	case vtrpcpb.Code_CLUSTER_EVENT:
+		isFailover = true
+	case vtrpcpb.Code_FAILED_PRECONDITION: // previous attempt, not used, to be removed
+		if strings.Contains(err.Error(), ClusterEventMoveTables) {
+			isFailover = true
+		}
+	}
```

Comment:
currently the `reason` is only used for logging. Will this be used anyplace else?

Reply:
Was only added for better visibility via logging. Maybe we can add a metric with this in the future if requested ...